### ISSUE TITLE
With jasmine-node, selenium/webdriverio hangs after third test

### ISSUE THREAD

**Issue by christopherbalz** [webdriverio/webdriverio#391]
Not sure this is a webdriverio problem, but if not perhaps there is some advice.  No matter what spec I run, how many specs I run, or what the test content is (`it('clears the search box')...`), after the third test, with jasmine-node running under webdriverio, the process hangs.  (Also posted on jasmine-node: https://github.com/mhevery/jasmine-node/issues/364 ).

Happens on Mac OS X Yosemite, up to date, and with all browsers.

With the `--forceexit` option, tests that pass when there are less than three total 'spec' files give this: `  timeout: timed out after 30000 msec waiting for spec to complete`.

Sometimes, upon cntrl-c, I get an error message about failing to connect, but I think that is caused by the blocked tests running after Selenium is killed:

```
### Execution Started ### Running Test xyz
{ [RuntimeError: RuntimeError
     (ECONNREFUSED:-1) Couldn't connect to selenium server
     Problem: Couldn't connect to selenium server

     Callstack:
     -> init()
]
  message: 'RuntimeError\n     (ECONNREFUSED:-1) Couldn\'t connect to selenium server\n     Problem: Couldn\'t connect to selenium server\u001b[1;30m\n\n     Callstack:\n     -> init()\n',
  name: 'RuntimeError',
  hasCallStack: true }

-------END TEST RESULTS-------
```

Webdriver is started in this case normally:

```
    beforeEach(function() {
        client = require('node_modules/webdriverio').remote({
        	port: process.env.seleniumPort || 4445,
            desiredCapabilities: {
                browserName: process.env.browser || 'safari'
            }
        }).init().url(<a-string-url>);
                console.log('\n### Execution Started ### Running Test which ' + jasmine.getEnv().currentSpec.description);
    }, TEST_TIMEOUT);
```

'jasmine.getEnv()' works fine, and so forth.

**Comment by hellatan** [comment]
@christopherbalz are you ending your specs with `.call(done)` per the examples? What's your timeout set to?

**Comment by christopherbalz** [comment]
Yes all is done per the example in the fail case.  Timeout is 30000 (ms).  The end of the 'it' blocks looks like this: 

```
.call(done);
}, TEST_TIMEOUT);
```

There is gitter chat dialog on this issue from Friday evening and Saturday: https://gitter.im/webdriverio/webdriverio?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge

**Comment by christopherbalz** [comment]
We found that the issue is not present when we do not start selenium from inside gulp. I don't know why this is, but we do have all the tests running now. I will update the git issues with the same. The tests currently follow the example webdriver.io.with.jasmine.spec, doing the beforeEach and afterEach as in that example (I could not get it to work otherwise, myself). And yes, I have had a great experience with webdriverio and mocha+chai as early as two years ago. Currently my team wanted to go with jasmine. Thanks!

**Comment by christopherbalz** [comment]
We found that the issue is not present when we do not start selenium from inside gulp. I don't know why this is, but we do have all the tests running now. I will update the git issues with the same. The tests currently follow the example webdriver.io.with.jasmine.spec, doing the beforeEach and afterEach as in that example (I could not get it to work otherwise, myself). And yes, I have had a great experience with webdriverio and mocha+chai as early as two years ago. Currently my team wanted to go with jasmine. Thanks!

**Comment by christopherbalz** [comment]
Actually the trigger for the issue was somewhere not in just starting selenium from gulp, which does work fine, but in combining that with running the tests in the same parent gulp task.

**Comment by christian-bromann** [comment]
mhm ok .. interesting

**Comment by christopherbalz** [comment]
Separating start-selenium, run-tests, and kill-selenium out as different gulp tasks still allowed us to run them all together from a single Makefile target.  So, that worked out pretty well, although I'm not sure what made the difference in not triggering the selenium/etc hang.

**Comment by binarykitchen** [comment]
Having the same issue here. See https://github.com/vvo/selenium-standalone/issues/77

How can I listen for session delete and end from a gulp task if it has no reference to the client?

**Comment by binarykitchen** [comment]
@christopherbalz do you have a working example with gulp?

**Comment by christopherbalz** [comment]
Yes - the key is what I stated above in my just-previous comment (unfortunately, I cannot share the code itself as the code is proprietary).   As for monitoring selenium from gulp (if that is what is being asked above), you can use the Selenium http api with nodejs `request` to see if it's running.  Or if you are asking how to shut down properly, this is a common way to do it: 

```
      selenium.on('close', function () {
        selenium.kill();
      });
```

**Comment by binarykitchen** [comment]
thanks

**Comment by nicsnoek** [comment]
I found my tests choking after a fixed number of webdriver steps. One way I found around this is to run selenium as a detached process, eg:

```javascript
var selenium = require(‘selenium-standalone’);
var fs = require('fs’);
var seleniumOut = fs.openSync(’./log/selenium.log’, 'a’);
selenium.start({
  spawnOptions: {
    detached: true,
    stdio: [ 'ignore’, seleniumOut, seleniumOut ]
  }
}
```

**Comment by jvilk** [comment]
@nicsnoek Thanks so much; I had the same exact issue as you. Specifying a log file fixes everything.

**Comment by lifeisfoo** [comment]
I'm facing the same issue using webdriverio in standalone mode, running it from mocha tests.

Since I have many tests, I started with a setup where the `init()` method was called before all tests, and `end()` after all. But after adding a test with many chained _isVisible_ method calls, some tests started to fail with the `ECONNREFUSED ` error.

I fixed this error, calling `init` and `end` methods just before and after each test case.
